Specification of flat serialisation for UPLC. (#2562)
input-output-hk/plutus: plutus-core-spec/plutus-core-specification.tex

> %g Plutus Core Specification

Author: Yes.

closure $C$ with $\discharge{C} = \erase{V}$.
\end{itemize}

\section{Serialisation}

> We use the \texttt{flat} format to serialise plutus terms. The \texttt{flat} format

Author: I am not sure this is in the spirit of the document. The feeling I got is that this is supposed to be an arid, concise specification of the format. I feel somewhat guilty even adding one example (maybe I should remove it and only keep the `Flat.md` file). Maybe I could add this motivation to the example, I think that would be best.

> We use the \texttt{flat} format to serialise plutus terms. The \texttt{flat} format

Reviewer: I think you could say something like: "The encoding we use follows the `flat` Haskell package (LINK), but this document provides an entirely standalone description of the format."

encodes sum types as tagged unions and products by concatenating their contents. We
proceed by defining the structure and the data types of untyped plutus core and how
they get serialised.

\subsection{Variable length data}

\texttt{Lists} are encoded by prefixing (using the most significant bit) the element
stored in the list with one of the two list constructor tags. If we want to follow the
current element with another one we use the bit `1` for the \texttt{Cons} constructor,
or `0` for the \texttt{Nil} constructor if this is the last element of the list.

\noindent We encode variable length data as a list of chunks, 7 bits each, with the least

> significant chunk first and the most significant bit first in the chunk.

Author: I have checked the serialisation output and everything seems fine. I wrote documentation of the steps I took to verify, so we can continue the discussion using that file instead of this one.

> significant chunk first and the most significant bit first in the chunk.

Author: After spending a bit of time unwinding the serialised format it looks like it uses one byte for constants that can be encoded in 7bits (do you think we should optimise this?). Maybe things will make more sense after you will read through the example serialisation of `(program 11.22.33 (con integer 11))`.

\medskip
\noindent Lets calculate the encoding of the \texttt{32768} index (unsigned, arbitrary
length integer):
\begin{enumerate}
  \item Converting \texttt{32768} to binary: \\
    \verb|32768| $\rightarrow$ \verb|0b1000000000000000|
  \item Split into 7 bit chunks: \\
    \verb|0b1000000000000000| $\rightarrow$ \verb|0000010 0000000 0000000|
  \item Reorder chunks (least significant chunk first): \\
    \verb|0000010 0000000 0000000| $\rightarrow$ \verb|0000000 0000000 0000010|
  \item Add list constructor tags: \\
    \verb|0000000 0000000 0000010| $\rightarrow$ \verb|10000000 10000000 00000010|
\end{enumerate}

\noindent Encoding primtive types:

\begin{itemize}
  \item Unsigned \texttt{Integer}s and \texttt{ByteString}s are encoded using the
    previously introduced encoding for variable length data types.
  \item Signed \texttt{Integer}s are first converted to an unsigned value using the

>     \texttt{ZigZag} encoding, then they are encoded as variable length data types.

Author: There is no reference I could find, but it is a well-known encoding. I added a footnote that describes why `ZigZag` is generally used, and how the encoding is done.

  \hline
  ByteString & 1 & Variable length \rule{0mm}{4mm} \\[\sep]
  Integer & 0 & ZigZag + Variable length \\[\sep]
  String & 2 & UTF-8 \\[\sep]
  Char & 3 & UTF-8 \\[\sep]
  Unit & 4 & Empty \\[\sep]
  Bool & 5 & `1` is True, `0` is False \\[\sep]
  \hline
\end{tabular}
\captionof{figure}{Constant kinds}
\label{fig:serialisation-constants}
\end{minipage}

\subsection{Untyped terms}

Terms are encoded using 4 bits, which allows a total of 16 terms, of which 8 are
currently used.

\vspace{1cm}

\begin{minipage}{\linewidth}
\centering
\begin{tabular}{|l|c|l|}
  \hline
  \Strut
  \textrm{Name} & \textrm{Tag} & \textrm{Arguments} \\
  \hline
  Variable & 0 & name \rule{0mm}{4mm} \\[\sep]
  Delay & 1 & term \\[\sep]
  Lambda abstraction & 2 & name, term \\[\sep]
  Application & 3 & term, term \\[\sep]
  Constant & 4 & constant \\[\sep]
  Force & 5 & term \\[\sep]
  Error & 6 & term \\[\sep]
  Builtin & 7 & builtin \\[\sep]
  \hline
\end{tabular}
\captionof{figure}{Untyped terms}
\label{fig:serialisation-terms}
\end{minipage}

\vspace{1cm}

\noindent Builtins use 5 bits for their definition, allowing for a maximum of 32 builtin
functions of which 25 are used.

> 

Reviewer 2: Having maximum of 32 builtin functions seems like a really, really bad idea. When we add bit ops for `Integer`s, `String` ops, not even saying about other crypto primitives, – and we are way over 32.

Author: This document is in sync with the code that we have right now, but realistically these values will change a lot as we get closer to the release. I am not sure if this is the best place to decide the number of builtins we will support.

My rule of thumb is to use a size that can accommodate we have right now.

Author: However, I don't really have a problem with increasing the number of bits at this point either. @nau do you have any suggestion?

Reviewer 2: I'm just worrying that these sizes could be hard to change afterwards. It's good to know those are not set in stone yet. I'd discuss and establish an amount of _reserved_ builtins we should remain for the future, and state it explicitly. Like, we'd like to have, say, at least 32 reserved builtins or something.

Reviewer: Increasing this limit would require changing the binary format of scripts. However, this is not so bad, because *most* changes to the language will require a "new" language, and thus it's relatively easy to say it has a new format.

It's *nice* to keep the format the same as much as possible, but it's not critical. (This reasoning should get written down somewhere...)

Reviewer: That said... I'm sure flat has a way to encode arbitrary-length integers. It might be worth seeing how much space we use if we encode the builtin tags as arbitrary-length integers instead. If it's cheap we might as well do it, I guess.

Author: It uses 1 byte to encode a 7bit value, and it uses minimum 1 byte.

Reviewer 3: Having thought about this some more, I'm wondering whether we should allow a variable-length encoding for built-in types and functions, like the one that you can derive automatically using `Generic`.  The automatically-generated encoding uses a prefix code (I think it's actually a truncated binary encoding, a Huffman encoding with uniform weights: see https://en.wikipedia.org/wiki/Truncated_binary_encoding).  Since it's a prefix code, the decoder can tell when it's got to the end of an encoding without having to know the length in advance.  

The reason I'm suggesting this is that the plan is to allow users to implement their own builtin types and functions, and (a) there's no reason why these should bear any relation to the builtins we have here, and (b) there may be more types/functions than you can fit into a predetermined number of bits.  With `flat`'s generic encoding we could just say that when you implement your builtins you use that specific encoding (perhaps implemented by hand if you're not using Haskell) and there would be no necessity to fix the size of the encoding in advance.  We should probably postpone that until another PR though.  At some point we're going to have to specify the mechanism for extensible builtins (SCP_1058) and that might be a good time to think about this again (ie, put it off for as long as possible).

By the way, is there any guarantee that `Generic` will always do the same thing?  Is there any danger that if you recompile your program in 10 years, it'll process constructors in a different way or something?   If that's a possibility it would mean that we definitely shouldn't use generically-generated encodings.

\vspace{1cm}

\begin{minipage}{\linewidth}
\centering
\begin{tabular}{|c|c|c|c|c|c|}
  \hline
  \Strut
  \textrm{Name} & \textrm{Tag} & \textrm{Name} & \textrm{Tag} & \textrm{Name} & \textrm{Tag} \\
  \hline
   AddInteger & 0 & GreaterThanEqInteger & 8 & QuotientInteger & 17 \rule{0mm}{4mm} \\[\sep]
   SubtractInteger & 1 & EqInteger & 9 & ModInteger & 18 \\[\sep]
   MultiplyInteger & 2 & Concatenate & 10 & LtByteString & 19 \\[\sep]
   DivideInteger & 3 & TakeByteString & 11 & GtByteString & 20 \\[\sep]
   RemainderInteger & 4 & DropByteString & 12 & IfThenElse & 21 \\[\sep]
   LessThanInteger & 5 & SHA2 & 13 & CharToString & 22 \\[\sep]
   LessThanEqInteger & 6 & SHA3 & 14 & Append & 23 \\[\sep]
   GreaterThanInteger & 7 & VerifySignature & 15 & Trace & 24 \\[\sep]
   & & EqByteString & 16 & & \\[\sep]
   \hline
\end{tabular}
\captionof{figure}{Builtin tags}
\label{fig:serialisation-builtins}
\end{minipage}

\vspace{1cm}

> 

Author: Do you think I should include the `Program` definition in the specification?

> 

Reviewer: There's an interesting question here. What people often do is have a more-stable "container" format that is basically just a version tag so you can tell how to interpret the body. We could do the same here, and specify the version header as the first few bytes of the file, and state that this is intended to be stable even if the body encoding changes later.

I like this, it would let us be fairly confident that we could always deserialize PLC files without having to know what version they are beforehand.

> 

Author: We always encode a `Program` which does contain versioning information, so I think we are already doing that?

> 

Reviewer: Well:
a) We want to be *extra* sure that the outer encoding is stable, so it will *always* be three variable length integers in exactly the same format, no matter what else changes. So that's worth calling out separately.
b) It's easy for that to go wrong if, say, we ever make use of programs with non-unit annotations (which we might want for, say, debugging). Then the annotation for Programs would go before the version, oh no!

So I think we need a specification of the header separately, and maybe a custom serialization format for Programs that puts the version before the annotation (and maybe remove the annotation for `Version`, too).

> 

Reviewer 3: I was wondering how you'd deal with scripts that are using non-standard builtins, since then the encoding of a builtin could refer to something completely unexpected.  I think we discussed this before and decided that it was just up to users to make sure they were doing the right thing.  Now I'm wondering if we could generate a hash form the names of the types and the signatures of the builtin functions and stick it in the header somewhere.

closure $C$ with $\discharge{C} = \erase{V}$.
\end{itemize}

\section{Serialisation}

> We use the \texttt{flat} \citep{flat} format to serialise plutus terms. The

Reviewer: ```suggestion
We use the \texttt{flat} \citep{flat} format to serialise Plutus Core terms. The
```

\texttt{flat} format encodes sum types as tagged unions and products by
concatenating their contents. We proceed by defining the structure and the data

> types of untyped plutus core and how they get serialised.

Reviewer: ```suggestion
types of untyped Plutus Core and how they get serialised.
```

\subsection{Variable length data}

\texttt{Non-empty lists} are encoded by prefixing the element stored with `0'
if this is the last element or `1' if there is more data following.

\noindent We encode variable length data as a non-empty list of chunks, 7 bits each,
with the least significant chunk first and the most significant bit first in the chunk.

\medskip

> \noindent Lets calculate the encoding of the \texttt{32768} index (unsigned, arbitrary

Reviewer: ```suggestion
\noindent Let's calculate the encoding of the \texttt{32768} index (unsigned, arbitrary
```

length integer):
\begin{enumerate}
  \item Converting \texttt{32768} to binary: \\
    \verb|32768| $\rightarrow$ \verb|0b1000000000000000|
  \item Split into 7 bit chunks: \\
    \verb|0b1000000000000000| $\rightarrow$ \verb|0000010 0000000 0000000|

>   \item Reorder chunks (least significant chunk first): \\

Reviewer: Missing the bit ordering step after this?

Author: I think I should remove this example, now that I will add the more comprehensive one.

Author: There is also, no reordering bits after this. Bits are already properly ordered, this would be the final binary output.

    \verb|0000010 0000000 0000000| $\rightarrow$ \verb|0000000 0000000 0000010|
  \item Add list constructor tags: \\
    \verb|0000000 0000000 0000010| $\rightarrow$ \verb|10000000 10000000 00000010|
\end{enumerate}

\subsection{Constants}

> Constants are encoded as a combination of the kind of value that is serialized and the

Reviewer: We typically refer to the value that indicates the kind as a "tag", let's use that throughout.
```suggestion
Constants are encoded as a combination of a tag indicating the kind of value that is serialized and the
```

> value itself. Constants use 4 bits to encode the constructors, so they allow for a

Reviewer: ```suggestion
value itself. Constants use 4 bits to encode the tags, so they allow for a
```

> maximum of 16 constructors, of which 6 are currently used.

Reviewer: ```suggestion
maximum of 16 constructors, of which 6 are used by the default set of builtin types.
```

\vspace{1cm}

\begin{minipage}{\linewidth}
\centering
\begin{tabular}{|l|c|l|}
  \hline
  \Strut
  \textrm{Name} & \textrm{Tag} & \textrm{Encoding} \\
  \hline
  ByteString & 1 & Variable length \rule{0mm}{4mm} \\[\sep]

>   Integer & 0 & ZigZag + Variable length \\[\sep]

Reviewer: Why not order them by their tag?

  String & 2 & UTF-8 \\[\sep]
  Char & 3 & UTF-8 \\[\sep]
  Unit & 4 & Empty \\[\sep]
  Bool & 5 & `1' is True, `0' is False \\[\sep]
  \hline
\end{tabular}
\captionof{figure}{Constant kinds}
\label{fig:serialisation-constants}
\end{minipage}

\subsection{Untyped terms}

> Terms are encoded using 4 bits, which allows a total of 16 terms, of which 8 are

Reviewer: ```suggestion
Terms are encoded using 4 bit tags, which allows a total of 16 kinds of term, of which 8 are
```

currently used.

\vspace{1cm}

\begin{minipage}{\linewidth}
\centering
\begin{tabular}{|l|c|l|}
  \hline
  \Strut
  \textrm{Name} & \textrm{Tag} & \textrm{Arguments} \\
  \hline
  Variable & 0 & name \rule{0mm}{4mm} \\[\sep]
  Delay & 1 & term \\[\sep]
  Lambda abstraction & 2 & name, term \\[\sep]
  Application & 3 & term, term \\[\sep]
  Constant & 4 & constant \\[\sep]
  Force & 5 & term \\[\sep]
  Error & 6 & term \\[\sep]
  Builtin & 7 & builtin \\[\sep]
  \hline
\end{tabular}
\captionof{figure}{Untyped terms}
\label{fig:serialisation-terms}
\end{minipage}

\vspace{1cm}

> \noindent Builtins use 5 bits for their definition, allowing for a maximum of 32 builtin

Reviewer: ```suggestion
\noindent Builtins use 5 bits for their tags, allowing for a maximum of 32 builtin
```

functions of which 25 are used.

\vspace{1cm}

\begin{minipage}{\linewidth}
\centering
\begin{tabular}{|c|c|c|c|c|c|}
  \hline
  \Strut
  \textrm{Name} & \textrm{Tag} & \textrm{Name} & \textrm{Tag} & \textrm{Name} & \textrm{Tag} \\
  \hline
   AddInteger & 0 & GreaterThanEqInteger & 8 & QuotientInteger & 17 \rule{0mm}{4mm} \\[\sep]
   SubtractInteger & 1 & EqInteger & 9 & ModInteger & 18 \\[\sep]
   MultiplyInteger & 2 & Concatenate & 10 & LtByteString & 19 \\[\sep]
   DivideInteger & 3 & TakeByteString & 11 & GtByteString & 20 \\[\sep]
   RemainderInteger & 4 & DropByteString & 12 & IfThenElse & 21 \\[\sep]
   LessThanInteger & 5 & SHA2 & 13 & CharToString & 22 \\[\sep]
   LessThanEqInteger & 6 & SHA3 & 14 & Append & 23 \\[\sep]
   GreaterThanInteger & 7 & VerifySignature & 15 & Trace & 24 \\[\sep]
   & & EqByteString & 16 & & \\[\sep]
   \hline
\end{tabular}
\captionof{figure}{Builtin tags}
\label{fig:serialisation-builtins}
\end{minipage}

\vspace{1cm}

> \noindent Encoding primtive types:

Reviewer: This is a little confusing in contrast to the section on encoding builtin types. I think the intention is to say that a *builtin* integer constant is encoded with a tag followed by a *primitive* integer, but maybe worth being extra clear about it.

Author: Changed `Encoding primitive types` to `Encoding primitive values`.

\begin{itemize}

>   \item Unsigned \texttt{Integer}s and \texttt{ByteString}s are encoded using the

Author: Very good catch. I was secretly hoping I don't have to talk about pre-aligned data structures. I think, at some point, it would be *very* useful to have an example encoding of some program that covers all Plutus Core data types, together with a test to check that the encoding and example are in sync. This would prevent changes to GHC Generic deriving mechanism affecting us. @michaelpj @kwxm, should I create an issue for it?

>   \item Unsigned \texttt{Integer}s and \texttt{ByteString}s are encoded using the

Author: I have update the specification with the encoding for `ByteString`, `Char` and `String`.

    previously introduced encoding for variable length data types.
  \item Signed \texttt{Integer}s are first converted to an unsigned value using the
    \texttt{ZigZag}\footnote{The \texttt{ZigZag} encoding interleaves positive and negative numbers such that small
    negative numbers are stored using a small number of bytes.} encoding, then they are encoded as variable length data types.
  \item \texttt{String}s are encoded as lists of characters and use the \texttt{UTF-8}
    encoding.
  \item \texttt{Unit} and data structures which have only one constructor, are removed
    from the serialised data.
  \item Variable names are encoded using DeBruijn indices, which are unsigned, arbitrary
    length integers.
\end{itemize}

> Possibly empty lists are encoded in the standard way, by the tag for the constructor,

Reviewer: You already talk about lists earlier.

Author: Those were `non-empty` lists, I think.
